src/chain_sync/chain_muxer.rs

```diff
@@ -454,8 +402,13 @@ where
                         return Ok(None);
                     }
                     // Assemble full tipset from block only in stateful mode
-                    let tipset =
-                        Self::gossipsub_block_to_full_tipset(b, source, network.clone()).await?;
+                    let tipset = Self::get_full_tipset(
+                        network.clone(),
+                        chain_store.clone(),
+                        source,
```

Comment:
In the previous logic, bitswap reqeusts are sent to all connected peers, should we fallback to non-source peers if it fails here?

Reply:
No, if a peer tells us about a new block without having it themselves, we should ignore it.

Reply:
Oh wait, as I understand, in gossipsub source could a proxy peer that does not actually own the block

Reply:
https://docs.rs/libp2p-gossipsub/latest/libp2p_gossipsub/enum.Event.html#variants

> propagation_source: [PeerId](https://docs.rs/libp2p-identity/0.2.9/x86_64-unknown-linux-gnu/libp2p_identity/peer_id/struct.PeerId.html)
The peer that forwarded us this message.

Reply:
"Fixed" by broadcasting the chain_exchange request. Eventually, all of this code needs to be reworked.